jgrapht-core/src/main/java/org/jgrapht/alg/flow/mincost/CapacityScalingMinimumCostFlow.java

```diff
@@ -211,40 +211,27 @@ public V getFlowDirection(E edge) {
      * It is represented as a mapping from graph nodes to their potentials (dual variables). Reduced cost
      * of a arc $(a, b)$ is defined as $cost((a, b)) + potential(b) - potential(b)$. According
      * to the reduced cost optimality conditions, a feasible solution to the minimum cost flow problem is
-     * optimal if and if reduced cost of every non-saturated arc is greater than or equal to $0$.
+     * optimal if and only if reduced cost of every non-saturated arc is greater than or equal to $0$.
      *
      * @return solution to the dual linear program formulated on the network.
      */
-    public DualSolution<V> getDualSolution() {
-        dualSolution = lazyComputeDualSolution();
-        return dualSolution;
-    }
+    public Map<V, Double> getDualSolution() {
 
-    /**
-     * Lazily computes solution to the dual linear program formulated on the flow network.
-     *
-     * @return a solution to the dual linear program
-     */
-    private DualSolution<V> lazyComputeDualSolution() {
-        lazyCalculateMinimumCostFlow();
-        if (dualSolution == null) {
-            Map<V, Double> dualVariables = new HashMap<>();
-            for (int i = 0; i < n; i++) {
-                dualVariables.put(graphVertices.get(i), nodes[i].potential);
-            }
-            dualSolution = new DualSolution<>(dualVariables);
+        if(minimumCostFLow == null)
+            throw new RuntimeException("Cannot return a dual solution before getMinimumCostFlow(MinimumCostFlowProblem minimumCostFlowProblem) is invoked!");
```

Comment:
This is too generic to throw. Either throw some standard exception that makes sense here, or maybe create some custom exception (which extends `RuntimeException`). 

As a more general solution (which arguably would be used also in other places) we could have a generic `JGraphTException` which extends RuntimeException and all others custom exceptions extend this one. For example we have something like `NegativeCycleDetectedException`. @jsichi Opinions?

Reply:
IllegalStateException seems appropriate here.

I don't see much value in a JGraphTException for the kind of unchecked exceptions we currently throw.  It might make sense in the future if we ever introduce checked exceptions.

Reply:
Actually, instead of throwing an exception, we might as well return null?
@return solution to the dual linear program formulated on the network, or null if no Minimum Cost Flow Problem has been solved.

Would you agree?

Reply:
I try to avoid returning `null` these days. A similar solution is to return `Optional` but we are not using such a pattern in the library. Not sure whether we want to. @jsichi @Toptachamann Preferences?

Reply:
I think it's fine if we start introducing Optional.

Reply:
I really dislike using Optional here. Normally the user would invoke:
`Map<.> dualSol=getDual();`
Only when the user forgets to invoke `calculateMinimumCostFlow(.)` this method goes wrong.
However, by returning an optional, the simple `Map<.> dualSol=getDual();` now becomes:
```
Optional<Map<.>> optionalDualSol=getDual();
if(optionalDualSol.isPresent())
  dualSol=optionalDualSol.get();
```
This is so incredibly verbose, especially since in 99.9% of the cases this is not needed at all.

So I would strongly prefer to simply return null, which seems appropriate here, and let the user do an old-fashioned null check if necessary, or throw an IllegalStateException. @d-michail @jsichi what's your preference? I'd like to wrap this one up such that we can proceed with the release.

Reply:
No strong preference. The documentation should contain the necessary information.

Reply:
Ok, let's keep it as-is for now. @d-michail if satisfied, could you go ahead and merge?